src/Controls/src/Core/DependencyService.cs

```diff
@@ -83,16 +81,22 @@ public static class DependencyService
 		{
 			Type targetType = typeof(T);
 			Type implementorType = typeof(T);
-			if (!DependencyTypes.Contains(targetType))
-				DependencyTypes.Add(targetType);
+			AddDependencyTypeIfNeeded(targetType);
 
 			lock (s_dependencyLock)
 				DependencyImplementations[targetType] = new DependencyData { ImplementorType = implementorType, GlobalInstance = instance };
 		}
 
+		static void AddDependencyTypeIfNeeded(
+			[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)] Type type)
+		{
+			if (!DependencyTypes.Any(t => t.Type == type))
+				DependencyTypes.Add(new DependencyType(type));
+		}
+
 		[return: DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)]
 		static Type FindImplementor(Type target) =>
-			DependencyTypes.FirstOrDefault(t => target.IsAssignableFrom(t));
+			DependencyTypes.FirstOrDefault(t => target.IsAssignableFrom(t.Type))?.Type;
```

Comment:
Is it possible to create an extra `static` method here instead of a new `DependencyType` class?

```csharp
static bool IsAssignableFrom([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)] Type t) => t.IsAssignableFrom(t);
```
Or can you put it as:
```csharp
static Type FindImplementor([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)] Type target)
//...
```

Either of these would maybe solve the warning with less code changes.

Reply:
In this case the problem wasn't a missing annotation on the `Type target` argument but instead on the items of the `List<Type> DependencyTypes`. Ideally, we'd like to have something like `List<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)] Type>` but that doesn't work and as a workaround we need to introduce the wrapper type `DependencyType` which stores the annotation on the property.

I think I can simplify the `DependencyType` though. I added the `IEquatable<DependencyType>` implementation to make the `DependencyTypes.Contains` method work but I later replaced it with `Any` and the extra code isn't needed anymore.

Reply:
You could also change it to struct to avoid the allocations.

Reply:
@jonathanpeppers I think the fix for this warning is now as short as it can get 😄